mirgecom/simutil.py

```diff
@@ -104,46 +104,94 @@ def exact_sim_checkpoint(discr, exact_soln, visualizer, eos, q,
     cv = split_conserved(discr.dim, q)
     dependent_vars = eos.dependent_vars(cv)
-    expected_state = exact_soln(t=t, x_vec=nodes, eos=eos)
+
+    rank = 0
+    if comm is not None:
+        rank = comm.Get_rank()
+
+    maxerr = 0.0
+    if exact_soln is not None:
+        if profiler is not None:
+            profiler.StartTimer(f"checkpoint_soln_{step}")
+        actx = cv.mass.array_context
+        nodes = thaw(actx, discr.nodes())
+        expected_state = exact_soln(t=t, x_vec=nodes, eos=eos)
+        exp_resid = q - expected_state
+        err_norms = [discr.norm(v, np.inf) for v in exp_resid]
+        maxerr = max(err_norms)
+        if profiler is not None:
+            profiler.EndTimer(f"checkpoint_soln_{step}")
 
     if do_viz:
-        from mirgecom.euler import split_conserved
+        if profiler is not None:
+            profiler.StartTimer(f"checkpoint_viz_{step}")
         io_fields = [
-            ("cv", split_conserved(discr.dim, q)),
-            ("dv", dependent_vars),
-            ("exact_soln", expected_state),
-            ("residual", q - expected_state)
+            ("cv", cv),
+            ("dv", dependent_vars)
         ]
+        if exact_soln is not None:
+            exact_list = [
+                ("exact_soln", expected_state),
+                ("residual", exp_resid)
+            ]
+            io_fields.extend(exact_list)
+
         from mirgecom.io import make_rank_fname, make_par_fname
         rank_fn = make_rank_fname(basename=vizname, rank=rank, step=step, t=t)
         visualizer.write_parallel_vtk_file(
             comm, rank_fn, io_fields, overwrite=overwrite,
             par_manifest_filename=make_par_fname(basename=vizname, step=step, t=t))
+        if profiler is not None:
+            profiler.EndTimer(f"checkpoint_viz_{step}")
 
     if do_status is True:
         #        if constant_cfl is False:
         #            current_cfl = get_inviscid_cfl(discr=discr, q=q,
         #                                           eos=eos, dt=dt)
         statusmesg = make_status_message(discr=discr, t=t, step=step, dt=dt,
                                          cfl=cfl, dependent_vars=dependent_vars)
-        err = q-expected_state
-        err_norms = [discr.norm(v, np.inf) for v in err]
-        statusmesg += (
-            "\n------- errors="
-            + ", ".join("%.3g" % en for en in err_norms))
+        if exact_soln is not None:
+            statusmesg += (
+                "\n------- errors="
+                + ", ".join("%.3g" % en for en in err_norms))
+
         if rank == 0:
             logger.info(statusmesg)
 
-        maxerr = max(err_norms)
-        if maxerr > exittol:
-            raise ExactSolutionMismatch(step, t=t, state=q)
+    if profiler is not None:
+        profiler.EndTimer(f"checkpoint_{step}")
+
+    if maxerr > exittol:
+        raise ExactSolutionMismatch(step, t=t, state=q)
+
+
+def create_parallel_grid(comm, grid_generator):
```

Comment:
Documentation?

Reply:
Added - but possibly all messed up, heh.  We'll see.

Reply:
I actually think it is OK, except I didn't figure out how to document the MPI communicator in a great way.

Reply:
See 83bb205.